### DISCUSSION TITLE
Endpoint to stream binary output from URL

### DISCUSSION THREAD

**Discussion by xiki808** [fastapi/fastapi#8581] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@endpoint.get("/stream/{path:path}")
async def stream(path: str):
    response = requests.get("http://10.0.3.4:5000/api/bedroom?h=432", stream=True)
    return StreamingResponse(response.content, media_type="multipart/x-mixed-replace; boundary=frame")
```

### Description

I want to use Fast API to create an endpoint that will async read a stream from a URL and return it. The idea is to have streams such as an RTSP stream available from the Fast API project. 

Above in the code is one way I have tried to achieve this, but this is a bit out of my hands and I would appreciate any help. Most of the examples found are streams from files on disk.

### Operating System

Linux

### FastAPI Version

0.73.0

### Python Version

3.9

**Comment by xiki808** [comment]
```
@endpoint.get("/video/{path:path}")
async def getVideo(path: str):
    return StreamingResponse(services.receiveFromURL(path), media_type="video/mp4")


@endpoint.get("/image/{path:path}")
async def getImage(path: str):
    return StreamingResponse(services.receiveFromURL(path), media_type="image/jpeg")


@endpoint.get("/stream/{path:path}")
async def getStream(path: str):
    return StreamingResponse(services.receiveFromURL(path), media_type="multipart/x-mixed-replace; boundary=frame")



# Somwhere else in the "services" class....

....

def receiveFromURL(self, path):
        with urllib.request.urlopen(f"http://{self.ip}:{self.port}/{path}") as resp:
            yield from resp

...
````


I manage to get it working, I needed to download files and/or stream output from URLs from my other services and this solution does the job. 

It is slower thought, especially for the video download. For example a clip of around 754kB takes ~2.5s instead of ~500ms. It gets quickly slower if I have the streams going on at the same time.

Would appreciate any feedback and improvements. Thank you!

**Comment by rafsaf** [comment]
Both `urllib` and `requests` are synchronous, you block event loop using them in `async` endpoints. It means that if request to external site tooks 10s, and you make 2 requests for the endpoint, they will took 20s instead of 10s.

In general [python-httpx](https://www.python-httpx.org/) fits better for async world of FastAPI (you can use httpx for async testing your endpoints also). And... they have an example of forwarding `StreamingResponse` implementation you ask, for Starlette ;) 
See [their docs here](https://www.python-httpx.org/async/#streaming-responses).

**Comment by xiki808** [comment, marked as answer]
Thank you @rafsaf for the information!

This is my updated generators, split in two, for those continuous streams and for the file download streams:

```
async def receiveFile(self, path):
        async with AsyncClient() as client:
            response = await client.get(f"http://{self.ip}:{self.port}/{path}")

            async for dataBytes in response.aiter_bytes():
                yield dataBytes

    async def receiveStream(self, path):
        async with AsyncClient() as client:
            async with client.stream('GET', f"http://{self.ip}:{self.port}/{path}") as response:
                async for dataBytes in response.aiter_bytes():
                    yield dataBytes
```

One thing that is of concern is that on every rerender of the component from the client side ( which is requesting the continuous stream ) a new process is created and it seems that the opened ones are never closed, not even when the browser is closed. Not sure if I can do something from the server side, but at the moment I'm investigating how to close it from the client side.

The processes I'm referring to are those triggered by `client.stream()`, which is a process from a service running in a docker container.
